cmd/kubeadm/app/phases/addons/dns/manifests.go

```diff
@@ -245,6 +248,15 @@ spec:
         effect: NoSchedule
       nodeSelector:
         beta.kubernetes.io/os: linux
+      affinity:
+        podAntiAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
+          - labelSelector:
+              matchExpressions:
+              - key: k8s-app
+                operator: In
+                values: ["kube-dns"]
+            topologyKey: kubernetes.io/hostname
```

Comment:
@rajansandeep could you please explain the motivation?

my understanding is the following:
- we reduce the replica count to 1.
- coredns will deploy on the primary CP node (where kubeadm init is called).
- the anti-affinity rule makes sure that the Pod will not schedule on a Node that already has it.

if i'm not mistaken, this will not improve much what we have right now.
a problem we have currently, is that both replicas land on the same primary CP Node.

ideally what we want is a coredns instance to be deployed on all CP Nodes.
one way of doing that is with static-pods, but given we treat coredns as an addon we should use a DaemonSet with a NodeSelector that matches the kubeadm "master" node-role.

i'm going to experiment with that in a bit.

Reply:
sadly, by changing the coredns object type we are going to break a lot of users that have automation around `kubectl patch deployment coredns...`.

Reply:
@neolit123 
With pod anti-affinity enabled and 2 coredns replicas:

- If a user has only a master node installed via `kubeadm init`, there will be one coredns pod in `running` state and one in `pending` state.
- The other coredns pod will remain in `pending` state and waits for scheduling until another worker node is created via `kubeadm join`.

Reply:
> a problem we have currently, is that both replicas land on the same primary CP Node.

Pod anti-affinity solves this problem.

Reply:
> With pod anti-affinity enabled and 2 coredns replicas:
> If a user has only a master node installed via kubeadm init, there will be one coredns pod in running state and one in pending state.
> The other coredns pod will remain in pending state and waits for scheduling until another worker node is created via kubeadm join.

this is true for 2 replicas, we don't want Pending pods because it will break e2e tests using our test suite, where all pods are expected to be ready.

>> a problem we have currently, is that both replicas land on the same primary CP Node.
> Pod anti-affinity solves this problem.

yes. but we reduce the replicas to 1, so if the primary CP node becomes NotReady (e.g. shutdown) the coredns service will go down too and the pod will not reschedule on a Ready node.

i guess i'm trying to see how 1 replica with anti-affinity is an improvement over 2 replicas without it.

like i've mentioned earlier, ideally we want a coredns DS for all CP nodes.

Reply:
if continuing to use a Deployment we might want to add these tolerations: https://github.com/kubernetes/kubernetes/issues/55713#issuecomment-538625309

Reply:
@chrisohaver PTAL too.

so basically i'm proposing that we keep the replica count to 2.
and introduce the following:
```
spec:
...
  tolerations:
  - key: "node.kubernetes.io/unreachable"
    operator: "Exists"
    effect: "NoExecute"
    tolerationSeconds: 15
  - key: "node.kubernetes.io/not-ready"
    operator: "Exists"
    effect: "NoExecute"
    tolerationSeconds: 15
```

this will improve the current deployment by rescheduling the coredns Pods if a Node becomes NotReady after 15 seconds.

i don't think the anti-affinity rule is needed here:
```
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: k8s-app
                operator: In
                values: ["kube-dns"]
            topologyKey: kubernetes.io/hostname
```

because with the current setup the deployment already does that.

Reply:
Let's start with the more trivial questions here. Is this required for the CoreDNS version bump?
If so, why is this a patch release and not a minor version bump? If it's not required, can we split it and move it into a separate PR?

Reply:
Yes, I've removed the pod anti-affinity changes from this PR and move it to another PR.